File: 2018/Day_01/code.php
<?php

require_once __DIR__ . '/../../lib/php/utils.php';

$input = readInput(__DIR__);

// Task code
function part01(array $input): int
{
    return array_reduce(
        $input,
        static function ($carry, $line) {
            return $carry + (int) $line;
        }
    );
}

function part02(array $input): int
{
    $freqList = [0 => 1];
    $freq = 0;

    while (true) {
        foreach ($input as $v) {
            $freq += $v;

            if ($freqList[$freq] ?? 0 === 1) {
                return $freq;
            }

            $freqList[$freq] = 1;
        }
    }
}

// Execute
calcExecutionTime();
$result01 = part01($input);
$result02 = part02($input);
$executionTime = calcExecutionTime();

writeln('Solution Part 1: ' . $result01);
writeln('Solution Part 2: ' . $result02);
writeln('Execution time: ' . $executionTime);

saveBenchmarkTime($executionTime, __DIR__);

// Task test
testResults(
    [574, 452], // Expected
    [$result01, $result02], // Result
);

File: 2021/Day_06/code.go
package main

import (
	"advent-of-code/helper"
	"fmt"
	"path"
	"runtime"
	"strconv"
	"strings"
	"time"
)

func main() {
	_, f, _, _ := runtime.Caller(0)
	cwd := path.Join(path.Dir(f))

	input := helper.ReadInput(cwd, helper.NewLine)
	iValues := input.Strings()

	// Execute
	start := time.Now()
	result01 := part01(iValues)
	result02 := part02(iValues)
	executionTime := helper.ExecutionTime(time.Since(start))

	fmt.Printf("Solution Part 1: %v\n", result01)
	fmt.Printf("Solution Part 2: %v\n", result02)
	fmt.Printf("Execution time: %s\n", executionTime)

	helper.SaveBenchmarkTime(executionTime, cwd)

	// Testing
	helper.TestResults(
		[]helper.TestingValue{
			helper.TestingValue{Result: result01, Expect: 385391},
			helper.TestingValue{Result: result02, Expect: 1728611055389},
		},
	)
}

// Task code
type Fish struct {
	days int
}

func (f *Fish) reproduce() *Fish {
	f.days--

	if f.days < 0 {
		fmt.Println(f.days)
		f.days = 6

		return &Fish{8}
	}

	return nil
}

func part01(input []string) int {
	return calcFishes(80, parsFishSchool(input[0]))
}

func part02(input []string) int {
	return calcFishes(256, parsFishSchool(input[0]))
}

func parsFishSchool(line string) map[int]int {
	fishes := make(map[int]int, 0)

	for _, s := range strings.Split(line, ",") {
		fish, _ := strconv.Atoi(s)
		fishes[fish]++
	}

	return fishes
}

func calcFishes(days int, startSchool map[int]int) int {
	fishSchool := startSchool
	for d := 1; d <= days; d++ {
		iteration := make(map[int]int, 0)

		for fd, f := range fishSchool {
			if fd == 0 {
				iteration[6] += f
				iteration[8] = f
			} else {
				iteration[fd-1] += f
			}
		}

		fishSchool = iteration
	}

	var count int
	for _, f := range fishSchool {
		count += f
	}

	return count
}
